src/line_simplication/visvalingam_whyatt.zig

```diff
@@ -0,0 +1,438 @@
+const HashPointAreaContext = struct {
+    /// Hashes the `index: usize` by bitcasting it to `u64`.
+    pub fn hash(_: HashPointAreaContext, merge_error: PointArea) u64 {
+        return @as(u64, @intCast(merge_error.index));
+    }
+    /// Compares two `index` for equality.
+    pub fn eql(_: HashPointAreaContext, merge_error_one: PointArea, merge_error_two: PointArea) bool {
+        return merge_error_one.index == merge_error_two.index;
+    }
+};
+
+/// Comparison function for the `HashedPriorityQueue`. It compares merge errors, and also considers
+///  bucket indices for equality.
+fn comparePointArea(_: void, point_1: PointArea, point_2: PointArea) math.Order {
+    if (point_1.area == point_2.area)
+        return math.Order.eq;
+    return math.order(point_1.area, point_2.area);
+}
+
+/// Return the absolute area of the triangle defined by three points.
+fn calculateArea(left_point: DiscretePoint, central_point: DiscretePoint, right_point: DiscretePoint) f64 {
+    const x1: f64 = @floatFromInt(left_point.time);
+    const y1: f64 = left_point.value;
+    const x2: f64 = @floatFromInt(central_point.time);
+    const y2: f64 = central_point.value;
+    const x3: f64 = @floatFromInt(right_point.time);
+    const y3: f64 = right_point.value;
+
+    return @abs((x1 * (y2 - y3) + x2 * (y3 - y1) + x3 * (y1 - y2)) / 2.0);
+}
+
+/// Append `value` of `type` determined at compile time to `compressed_values`.
+fn appendValue(comptime T: type, value: T, compressed_values: *std.ArrayList(u8)) !void {
+    // Compile-time type check
+    switch (@TypeOf(value)) {
+        f64, usize => {
+            const value_as_bytes: [8]u8 = @bitCast(value);
+            try compressed_values.appendSlice(value_as_bytes[0..]);
+        },
+        else => @compileError("Unsupported type for append value function"),
+    }
+}
+
+/// Creates a linear function that passes throught the two points of the `segment` and returns it
+/// in `linear_function`.
+fn createLinearFunction(start_point: DiscretePoint, end_point: DiscretePoint, linear_function: *LinearFunction) void {
+    if (end_point.time != start_point.time) {
+        const duration: f80 = @floatFromInt(end_point.time - start_point.time);
+        linear_function.slope = (end_point.value - start_point.value) / duration;
+        linear_function.intercept = start_point.value - linear_function.slope *
+            @as(f80, @floatFromInt(start_point.time));
+    } else {
+        linear_function.slope = 0.0;
+        linear_function.intercept = start_point.value;
+    }
+}
+
+/// Test if the area of all the triangles defined by three points contained in `values`
+/// is within the `error_bound`.
+pub fn testAreaWithinErrorBound(
+    values: []const f64,
+    error_bound: f32,
+) !void {
+    // At least three points are needed to form a triangle.
+    if (values.len < 3) return;
+
+    // Calculate the area formed by all triangles.
```

Comment:
Maybe "by each triangle" as it currently sounds like the total area of all are calculated.